src/accommodation-agent/Tools/AccommodationTools.cs

```diff
@@ -0,0 +1,139 @@
+using System.ComponentModel;
+using System.Text.Json;
+using Microsoft.Extensions.AI;
+using AccommodationAgent.Services;
+using AccommodationAgent.Models;
+
+namespace AccommodationAgent.Tools;
+
+public class AccommodationTools
+{
+    private readonly IAccommodationService _accommodationService;
+    private readonly IRerankingService _rerankingService;
+    private readonly IGeocodingService _geocodingService;
+    private readonly ILogger<AccommodationTools> _logger;
+
+    public AccommodationTools(
+        IAccommodationService accommodationService,
+        IRerankingService rerankingService,
+        IGeocodingService geocodingService,
+        ILogger<AccommodationTools> logger)
+    {
+        _accommodationService = accommodationService;
+        _rerankingService = rerankingService;
+        _geocodingService = geocodingService;
+        _logger = logger;
+    }
+
+    [Description("Geocode an address or landmark to get its coordinates (latitude, longitude)")]
+    public async Task<string> GeocodeLocationAsync(
+        [Description("Address or landmark name to geocode (e.g., 'Colosseum', 'Vatican', 'Rome', 'Latina')")] string location)
+    {
+        try
+        {
+            var coordinates = await _geocodingService.GeocodeAsync(location);
+            
+            if (coordinates.HasValue)
+            {
+                return JsonSerializer.Serialize(new
+                {
+                    location,
+                    latitude = coordinates.Value.Latitude,
+                    longitude = coordinates.Value.Longitude,
+                    message = $"Successfully geocoded '{location}' to coordinates."
+                });
+            }
+            
+            return JsonSerializer.Serialize(new
+            {
+                location,
+                message = $"Unable to geocode '{location}'. Location not found."
+            });
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, "Error geocoding location {Location}", location);
+            return JsonSerializer.Serialize(new { error = "An error occurred while geocoding the location." });
+        }
+    }
+
+    [Description("Search for accommodations based on various criteria including rating, location, amenities, price, and type")]
+    public async Task<string> SearchAccommodationsAsync(
+        [Description("The user's original search query to use for reranking results")] string userQuery,
+        [Description("Minimum user rating from 1 to 5 (e.g., 4 means at least 4 stars)")] double? minRating = null,
+        [Description("City name to search in (e.g., 'Rome', 'Latina')")] string? city = null,
```

Comment:
The city parameter is not required. Whenever the user asks for a city, it must be geocoded first with the Geocode tool, then coordinates will be passed here as input.